### ISSUE TITLE
learn FairScale 

### ISSUE THREAD

**Issue by vycezhong** [vycezhong/cs-notes#21]
https://github.com/facebookresearch/fairscale

> FairScale is a PyTorch extension library for high performance and large scale training. 

看上去是几个module的并集.

**Comment by vycezhong** [comment]
![image](https://user-images.githubusercontent.com/25879526/130314646-c78ccc69-b090-4d95-a4d0-1ebc92372925.png)

文档里这张图有点意思. 不过too slow的solution里面看上去有点奇怪，activation checkpoint, zero主要是节约memory的方法.

**Comment by vycezhong** [comment]
https://fairscale.readthedocs.io/en/latest/deep_dive/oss_sdp_fsdp.html

介绍了ZeRO.

## Optimizer State Sharding (OSS)

比如Adam做mixed precision training，还要保存一份FP32的momentum和variance. 这挺大的，而且每个worker都有一份，很多冗余. OSS的idea就是把这些FP32的optimizer states shard到各个worker上去，每个worker all-reduce gradients之后，只需要update自己那一份optimizer state，更新自己那一部分parameters，最后把自己这个shard对应的参数做一个broadcast/allgather. 

分析下额外的通信量，假如有N个worker，模型参数量为M，每个worker上的shard为M/N，因OSS的broadcast/allgather操作产生M/N * (N - 1)大小的通信量. 原本的all-reduce的通信量是 2 * M (N-1) / N. 所以OSS带来 1.5x 通信开销.

**Comment by vycezhong** [comment]
https://fairscale.readthedocs.io/en/latest/deep_dive/offload.html

介绍了ZeRO-offload. 具体见 https://github.com/vycezhong/read-papers/issues/193

看文档offload似乎有两种用法：
1. 一种就是ZeRO-offload，offload的是参数.
2. 另一外一种似乎是ZeRO activation offload，就是如果使用了activation checkpoint，可以offload checkpoint的activation.

**Comment by vycezhong** [comment]
https://fairscale.readthedocs.io/en/latest/deep_dive/adascale.html

## AdaScale

今年OSDI '21 best paper Pollux https://github.com/vycezhong/read-papers/issues/180 钦定了AdaScale https://github.com/vycezhong/read-papers/issues/149  

AdaScale是一种自动adapt lr的方法，适用于large batch和变batch size的情况. 原论文讨论的是SGD. 但Pollux似乎已经应用于Adam了. 所以文档提了一句: 

> This technique typically works well for SGD (with and without momentum) The assumption is that you already have a good learning rate schedule that works well for small batch sizes. (AdaScale has not been validated to work effectively with Adam, further research in that direction is needed.)

确实.

**Comment by vycezhong** [comment]
https://fairscale.readthedocs.io/en/latest/deep_dive/pipeline_parallelism.html

## Pipeline Parallelism

实现的是GPipe. 应该是来自https://github.com/kakaobrain/torchgpipe . 目前已经被整合进上游. https://pytorch.org/docs/stable/pipeline.html

**Comment by vycezhong** [comment]
https://fairscale.readthedocs.io/en/latest/deep_dive/activation_checkpointing.html

## Enhanced Activation Checkpointing

pytorch本身已经支持了activation checkpointing. fairscale enhance的地方是可以把这个checkpoint的activation offload到CPU，并可以handle forward函数的non-tensor输出（这有啥用？）

这个caveat不错，以前没注意到. 
> When using BatchNormalization you may need to freeze the calculation of statistics since we run the forward pass twice.

**Comment by vycezhong** [comment]
总结一下，fairscale可以说是large-scale training的缉书. 虽然代码基本是抄的，但fairscale可以说是总结了目前学界在large-scale training领域的重要成果. 

> fairscale.nn.pipe is forked from torchgpipe, Copyright 2019, Kakao Brain, licensed under Apache License.
> fairscale.nn.model_parallel is forked from Megatron-LM, Copyright 2020, NVIDIA CORPORATION, licensed under Apache License.
> fairscale.optim.adascale is forked from AdaptDL, Copyright 2020, Petuum, Inc., licensed under Apache License.
> fairscale.nn.misc.flatten_params_wrapper is forked from PyTorch-Reparam-Module, Copyright 2018, Tongzhou Wang, licensed under MIT License.

**Comment by vycezhong** [comment]
对ZeRO方法没什么兴趣. 先run pipeline parallelism. 打算跑一个8-stage的BERT-32.

**Comment by vycezhong** [comment]
暂时到此为止. 现在需要去深度了解Megatron-LM. 毕竟NVIDIA是真的用这个框架训练过大模型的！研究这个会更有意义一些.

**Comment by vycezhong** [comment]
准备看看MoE实现. 

https://github.com/facebookresearch/fairscale/blob/main/fairscale/nn/moe/moe_layer.py
